### PR TITLE
8780: Dedicated bitwise operators

### PR DESCRIPTION
**Pull request by jazzthief**
### Description

Resolves issue #8780. 
Support dedicated bitwise operators to provide uniform DBMS-agnostic access to them.

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [ ] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [x] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by zzzeek** [comment]
looks really good.   let me know when you bring it out of draft

**Comment by jazzthief** [comment]
Looking at it now, I think adding a column-expression-level function for `bitwise_not` overcomplicates things. I did it with the example of `distinct()` in mind, but reading the docs, where it is listed under "Column Element Foundational Constructors", I just don't see a simple bitwise NOT there.
I'll leave it  alone until we'll have someone else's opinion on this.

**Comment by jazzthief** [comment]
Also, Mike was talking about datatypes in the original issue #8780 (e.g. MySQL `BIT`): I haven't added any tests for that. If there should be more tests - just point me in the right direction.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **zzzeek** to try to get revision 379b1af4723e56abe7e7df7057658bdc62447666 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change 379b1af4723e56abe7e7df7057658bdc62447666: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4411

**Review by sqla-tester — COMMENTED** [review]
**Federico Caselli** (CaselIT) wrote:

Great work @jazzthief 

I'm not 100% convinced regarding the standalone bitwise_not. All other bitwise are available only on the columns, so it stands out.

Also it may confuse things a bit with "not_" (maybe not?)

mike what do you thin?

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4411

* lib/sqlalchemy/sql/_elements_constructors.py (line 1034): if we have this as a function I think we should mention in the docs that this should not be confused with the more common boolean `not_`
* test/sql/test_operators.py (line 4580): let's also add for each one a `lambda c1, c2: c1.bitwise_rshift_op(c2)` to ensure that the operator are properly used

**Comment by jazzthief** [comment]
@CaselIT Thanks for the review! 
I've addressed your comments. There is somehow no reference to standalone `bitwise_not()` in the built docs - looking into it.

**Comment by jazzthief** [comment]
Alright, now it is in the docs.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision a4541772a6a784f9161ad78ef84d2ea7a62fa8de of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset a4541772a6a784f9161ad78ef84d2ea7a62fa8de added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4411

**Comment by jazzthief** [comment]
@zzzeek 
> let's get those two in also

Those are done, patchset 2 on gerrit - unless there's something more we want to add to the docs to avoid confusion with `not_`.
Other than that, me and @CaselIT are not 100% sure we should have a standalone `bitwise_not()` - so it's your call :)

File: lib/sqlalchemy/sql/operators.py [hunk 1]
```diff
@@ -690,6 +690,36 @@ def ilike(
...
         return self.operate(ilike_op, other, escape=escape)
 
+    def bitwise_xor(self, other: Any) -> ColumnOperators:
+        """Return bitwise XOR operation"""
+
+        return self.operate(bitwise_xor_op, other)
+
+    def bitwise_or(self, other: Any) -> ColumnOperators:
+        """Return bitwise OR operation"""
+
```
Comment (sqla-tester) [inline lib/sqlalchemy/sql/operators.py:700]: **mike bayer** (zzzeek) wrote:

adding versionadded:: 2.0.2 to these

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4411

**Review by sqla-tester — COMMENTED** [review]
**mike bayer** (zzzeek) wrote:

doc updates, changelog, etc.   this is ready to go

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4411

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/4411 has been **merged**. Congratulations! :)

**Pull request closed without merging**